javaslang/src/main/java/javaslang/collection/PriorityQueue.java

```diff
@@ -0,0 +1,853 @@
+     * *     if rank t1 = rank t2 then skewLink(Node(x,0,[]),t1,t2) :: rest
+     * *     else                      Node (x,0,[]) :: ts
+     * * | insert (x, ts) =            Node (x,0,[]) :: ts
+     **/
+    public Q insert(T value) {
+        List<Node<Ranked<T>>> queue = this.forest;
+        Node<Ranked<T>> tree = Tree.of(Ranked.of(value, 0));
+        if (queue.size() >= 3) {
+            final Node<Ranked<T>> t1 = this.forest.head(), t2 = this.forest.get(1);
+            if (rank(t1) == rank(t2)) {
+                queue = this.forest.drop(2);
+                tree = skewLink(tree, t1, t2);
+            }
+        }
+        return with(List.of(tree).appendAll(queue));
+    }
+
+    /**
+     * fun meld (ts, ts') = meldUniq (uniqify ts, uniqify ts')
+     **/
+    public Q union(Q target) {
+        final List<Node<Ranked<T>>> forest = unionUnique(this.uniqify(), target.uniqify());
+        return with(forest);
+    }
+
+    /**
+     * fun findMin [] = raise EMPTY
+     * * | findMin [t] = root t
+     * * | findMin (t :: ts) =
+     * *     let val x = findMin ts
+     * *     in if Elem.leq (root t, x) then root t else x end
+     */
+    public T findMin() {
+        if (forest.isEmpty()) {
+            throw new IllegalStateException();
+        } else {
+            final T value = root(forest.head());
+            return (forest.size() == 1)
+                    ? value
+                    : min(value, with(forest.tail()).findMin());
+        }
+    }
+
+    private T min(T v1, T v2) {
+        return comparator.isLessOrEqual(v1, v2) ? v1 : v2;
+    }
+
+    /*
+    * fun deleteMin [] = raise EMPTY
+    * * | deleteMin ts =
+    * *     let fun getMin [t]=(t, [])
+    * *           | getMin (t :: ts) =
+    * *               let val (t', ts') = getMin ts
+    * *               in if Elem.leq(root t, root t') then (t, ts)
+    * *                  else                              (t', t :: ts') end
+    * *         fun split (ts,xs,[]) = (ts, xs)
+    * *           | split (ts,xs,t :: c) =
+    * *               if rank t = 0 then split (ts,root t :: xs,c)
+    * *               else               split (t :: ts,xs,c)
+    * *         val (Node (x,r,c), ts) = getMin ts
+    * *         val (ts',xs') = split ([],[],c)
+    * * in fold insert xs' (meld (ts, ts')) end
+    * */
+    @SuppressWarnings("Convert2MethodRef")
+    public Q deleteMin() {
+        if (forest.isEmpty()) {
+            throw new IllegalStateException();
+        } else {
+            final Tuple2<Node<Ranked<T>>, List<Node<Ranked<T>>>> min = with(forest).getMin();
+            final Tuple3<List<Node<Ranked<T>>>, List<T>, ?> split = split(List.empty(), List.empty(), min._1.getChildren());
+            final Q seed = with(min._2).union(with(split._1));
+            return split._2.foldLeft(seed, (ts, value) -> ts.insert(value));
+        }
+    }
+
+    Tuple2<Node<Ranked<T>>, List<Node<Ranked<T>>>> getMin() { // TODO stack overflow?
```

Comment:
Will investigate the depth of the recursion and whether it can cause `Stack overflow`